feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+	if val == nil {
+		return 0
+	}
+	return uint32(*val)
+}
+
+func GetPccRules(
+	policyDecision *n7_sbi.SmPolicyDecision,
+) ([]*protos.StaticRuleInstall, []*protos.DynamicRuleInstall, []string) {
+	var (
+		staticRules   []*protos.StaticRuleInstall
+		dynamicRules  []*protos.DynamicRuleInstall
+		rulesToRemove []string
+	)
+
+	if policyDecision.PccRules == nil {
+		return staticRules, dynamicRules, rulesToRemove
+	}
+
+	for ruleId, pccRule := range policyDecision.PccRules.AdditionalProperties {
+		if pccRule.PccRuleId == "" {
+			// This rule to be removed
+			rulesToRemove = append(rulesToRemove, ruleId)
+			continue
+		}
+		activationTime, deactivationTime := GetActivateDeactivationTime(policyDecision, &pccRule)
+		policyRule := GetPolicyRuleProto(policyDecision, &pccRule)
+		if policyRule == nil {
+			// This is a static rule. Static rule only has pccRuleId and RefCondData
+			staticRule := &protos.StaticRuleInstall{
+				RuleId:           ruleId,
+				ActivationTime:   activationTime,
+				DeactivationTime: deactivationTime,
+			}
+			staticRules = append(staticRules, staticRule)
+			continue
+		}
+
+		dynamicRule := &protos.DynamicRuleInstall{
+			PolicyRule:       policyRule,
+			ActivationTime:   activationTime,
+			DeactivationTime: deactivationTime,
+		}
+		dynamicRules = append(dynamicRules, dynamicRule)
+	}
+
+	return staticRules, dynamicRules, rulesToRemove
+}
+
+func GetActivateDeactivationTime(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) (activationTime, deactivationTime *timestamp.Timestamp) {
+	if pccRule.RefCondData != nil && *pccRule.RefCondData != "" {
+		condData, found := policyDecision.Conds.Get(*pccRule.RefCondData)
+		if found {
+			activationTime = ConvertToProtoTimeStamp(condData.ActivationTime)
+			deactivationTime = ConvertToProtoTimeStamp(condData.DeactivationTime)
+		}
+	}
+	return
+}
+
+func GetPolicyRuleProto(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.PolicyRule {
+	// For all Ref*Data Maximum one entry is expected.
+	// These are declared as arrays for future compatibility
+	policyRule := &protos.PolicyRule{
+		Id:            pccRule.PccRuleId,
+		Priority:      getSbiUinteger(pccRule.Precedence),
+		MonitoringKey: getMonKeyFromPccRule(policyDecision, pccRule),
+		Redirect:      getTcDataFromPccRule(policyDecision, pccRule),
+		Qos:           getQosDataFromPccRule(policyDecision, pccRule),
+		FlowList:      getFlowListFromPccRule(pccRule),
+	}
+	if pccRule.RefChgData != nil && len(*pccRule.RefChgData) > 0 && policyDecision.ChgDecs != nil {
+		chgData, found := policyDecision.ChgDecs.Get((*pccRule.RefChgData)[0])
+		if found {
+			policyRule.RatingGroup = uint32(*chgData.RatingGroup)
+			policyRule.ServiceIdentifier = &protos.ServiceIdentifier{
+				Value: uint32(*chgData.ServiceId),
+			}
+			policyRule.Offline = getSbiBool(chgData.Offline)
+			policyRule.Online = getSbiBool(chgData.Online)
+		}
+	}
```

Comment:
let's make this a function

```
func insertChargingAndMonitoringKeyIfNecessary(pccRule *n7_sbi.PccRule , policyRule *protos.PolicyRule){
      if pccRule.RefChgData != nil && len(*pccRule.RefChgData) > 0 && policyDecision.ChgDecs != nil {
		chgData, found := policyDecision.ChgDecs.Get((*pccRule.RefChgData)[0])
		if found {
			policyRule.RatingGroup = uint32(*chgData.RatingGroup)
			policyRule.ServiceIdentifier = &protos.ServiceIdentifier{
				Value: uint32(*chgData.ServiceId),
			}
			policyRule.Offline = getSbiBool(chgData.Offline)
			policyRule.Online = getSbiBool(chgData.Online)
		}
	}
```
see that `policyRule` object is a piointer so you will be able to modify it